src/crypto.h

```diff
@@ -43,6 +45,34 @@ struct PublicKey {
   int key_length;
 };
 
+class MultiPartSHA512Hasher {
+ public:
+  MultiPartSHA512Hasher() { crypto_hash_sha512_init(&state_); }
+  void update(const unsigned char *part, int64_t size) { crypto_hash_sha512_update(&state_, part, size); }
+  std::string getHexDigest() {
+    std::string sha512_hash(crypto_hash_sha512_BYTES, '\0');
+    crypto_hash_sha512_final(&state_, (unsigned char *)sha512_hash.c_str());
```

Comment:
This is safe because I allocate string with exact hash length(crypto_hash_sha512_BYTES) and crypto_hash_sha512_final writes exact bytes count. But If You don't like it I will rewrite it.